google-cloud-bigquery/lib/google/cloud/bigquery/schema.rb

```diff
@@ -123,6 +184,85 @@ def empty?
+        #
+        # The JSON schema file is the same as for the [`bq`
+        # CLI](https://cloud.google.com/bigquery/docs/schemas#specifying_a_json_schema_file)
+        # consisting of an array of JSON objects containing the following:
+        # - `name`: The column's [data
+        #   type](https://cloud.google.com/bigquery/docs/schemas#standard_sql_data_types)
+        # - `type`: The column [name](https://cloud.google.com/bigquery/docs/schemas#column_names)
+        # - `description`: (Optional) The column's [description](https://cloud.google.com/bigquery/docs/schemas#column_descriptions)
+        # - `mode`: (Optional) The column's [mode](https://cloud.google.com/bigquery/docs/schemas#modes)
+        #   (if unspecified, mode defaults to `NULLABLE`)
+        # - `fields`: If `type` is `RECORD`, an array of objects defining child
+        #   fields with these properties
+        #
+        # @param [IO, String, Array<Hash>] source An `IO` containing the JSON
+        #   schema, a `String` containing the JSON schema, or an `Array` of
+        #   `Hash`es containing the schema details.
+        #
+        # @return [Schema] The schema so that commands are chainable.
+        #
+        # @example
+        #   require "google/cloud/bigquery"
+        #
+        #   bigquery = Google::Cloud::Bigquery.new
+        #   dataset = bigquery.dataset "my_dataset"
+        #   table = dataset.table "my_table" do |table|
+        #     table.schema.load File.read("path/to/schema.json")
+        #   end
+        #
+        def load source
+          if source.respond_to?(:rewind) && source.respond_to?(:read)
+            source.rewind
+            schema_json = String source.read
+          elsif source.is_a? Array
+            schema_json = source.to_json
+          else
+            schema_json = String source
+          end
+
+          schema_json = migrate_json schema_json
+
+          raise ArgumentError, "Invalid JSON" unless schema_json
+
+          @gapi = Google::Apis::BigqueryV2::TableSchema.from_json schema_json
+
+          self
+        end
+
+        ##
+        # Write the schema as JSON to a file.
+        #
+        # The JSON schema file is the same as for the [`bq`
+        # CLI](https://cloud.google.com/bigquery/docs/schemas#specifying_a_json_schema_file).
+        #
+        # @param [IO, String] destination An `IO` to which to write the schema,
+        #   or a `String` containing the filename to write to.
+        #
+        # @return [Schema] The schema so that commands are chainable.
+        #
+        # @example
+        #   require "google/cloud/bigquery"
+        #
+        #   bigquery = Google::Cloud::Bigquery.new
+        #   dataset = bigquery.dataset "my_dataset"
+        #   table = dataset.table "my_table"
+        #   table.schema.dump "path/to/schema.json"
+        #
+        def dump destination
+          if destination.respond_to?(:rewind) && destination.respond_to?(:write)
```

Comment:
I bet if this code is changed to the following:

```ruby
if destination.respond_to?(:rewind) && destination.respond_to?(:write)
  destination.rewind
  destination.write JSON.dump(fields.map(&:to_hash))
else
  File.write String(destination), JSON.dump(fields.map(&:to_hash))
end
```

You can avoid the `*args` argument to `Field#to_json` (_and_ you can remove the `Field#to_json` method).